File: pkg-py/src/querychat/_handoff_server.py
from __future__ import annotations

import asyncio
import uuid
from typing import TYPE_CHECKING, Any

from shiny.types import NotifyException

from shiny import reactive, render, ui

from ._handoff_orchestrator import HandoffOrchestrator, parse_generate_payload

if TYPE_CHECKING:
    from collections.abc import Coroutine

    import chatlas
    import shinychat
    from shiny.bookmark import BookmarkState, RestoreState

    from shiny import Inputs, Session

    from ._datasource import DataSource
    from ._handoff_gallery import GalleryItem
    from ._handoff_orchestrator import GenerateRequest
    from ._handoff_prompt import Recommendation
    from ._query_executor import QueryExecutor


HANDOFFS_BOOKMARK_KEY = "querychat_handoffs"


def open_handoff_creator(
    orchestrator: HandoffOrchestrator,
    recommend_task: reactive.ExtendedTask[[list[GalleryItem]], Recommendation],
) -> None:
    items = orchestrator.open_modal()
    # Extended tasks queue repeated invocations, so discard a recommendation
    # started for a previous modal before requesting one for this gallery.
    recommend_task.cancel()
    if items:
        recommend_task.invoke(items)


async def set_active_handoff(
    orchestrator: HandoffOrchestrator,
    active_handoff_id: reactive.Value[str | None],
    handoff_id: str | None,
) -> None:
    active_handoff_id.set(handoff_id)
    await orchestrator.view.set_panel_open(is_open=handoff_id is not None)


def build_handoff_snapshot(
    orchestrator: HandoffOrchestrator,
) -> list[dict[str, Any]]:
    return orchestrator.store.bookmark_values()


def apply_handoff_snapshot(
    orchestrator: HandoffOrchestrator,
    values: object,
    active_handoff_id: reactive.Value[str | None],
) -> Coroutine[Any, Any, None] | None:
    if values is None:
        orchestrator.restore_snapshot([])
    elif isinstance(values, list):
        orchestrator.restore_snapshot(values)
    else:
        return None
    active_handoff_id.set(None)
    return orchestrator.view.set_panel_open(is_open=False)


def finish_handoff_restore_task(
    task: asyncio.Task[None],
    restore_tasks: set[asyncio.Task[None]],
) -> None:
    restore_tasks.discard(task)
    if task.cancelled():
        return
    try:
        task.result()
    except Exception as error:
        ui.notification_show(
            f"Failed to close the handoff panel after state restore: {error}",
            type="error",
            duration=None,
        )


async def save_handoff_revision(
    shinychat_chat: shinychat.Chat,
) -> None:
    await shinychat_chat.history.save()


async def generate_and_save_handoff(
    orchestrator: HandoffOrchestrator,
    request: GenerateRequest,
    directions: str,
    handoff_id: str,
    *,
    shinychat_chat: shinychat.Chat,
) -> None:
    await orchestrator.generate(request, directions, handoff_id)
    await save_handoff_revision(shinychat_chat)


def handoff_server(
    input: Inputs,
    session: Session,
    chat: chatlas.Chat,
    *,
    data_sources: dict[str, DataSource],
    executor: QueryExecutor,
    shinychat_chat: shinychat.Chat,
) -> None:
    orch = HandoffOrchestrator(
        session,
        chat,
        data_sources,
        executor,
        shinychat_chat,
    )
    active_handoff_id: reactive.Value[str | None] = reactive.Value(None)
    restore_tasks: set[asyncio.Task[None]] = set()

    @reactive.extended_task
    async def recommend_task(items: list[GalleryItem]) -> Recommendation:
        return await orch.recommend(items)

    @shinychat_chat.slash_command(
        "handoff",
        "Prepare a shareable handoff document or webapp using the current chat context.",
        echo=False,
    )
    async def open_handoff_modal():
        with reactive.isolate():
            stream_status = shinychat_chat.latest_message_stream.status()
        if stream_status == "running":
            await shinychat_chat.append_message(
                "Please wait for the current response to finish before preparing a handoff."
            )
            return

        open_handoff_creator(orch, recommend_task)

    @reactive.effect
    @reactive.event(recommend_task.status)
    async def on_recommend_complete():
        status = recommend_task.status()
        if status == "success":
            await orch.view.show_recommendation(recommend_task.result())
        elif status == "error":
            try:
                recommend_task.result()
            except Exception as e:
                error_msg = str(e)
            else:
                error_msg = "Unknown error"
            ui.notification_show(
                f"Auto-recommend failed: {error_msg}",
                type="error",
                duration=None,
            )
            await orch.view.show_recommendation_error(error_msg)

    @reactive.effect
    @reactive.event(input.handoff_generate)
    async def on_generate():
        req = parse_generate_payload(input.handoff_generate(), orch.default_type_id)
        if req.type_id == "other" and not req.freeform:
            ui.notification_show(
                "Please enter a format name for 'Other'.",
                type="warning",
            )
            return
        try:
            directions = input.handoff_directions() or ""
        except Exception:
            directions = ""

        # Open the panel before generation yields its first streamed source chunk.
        handoff_id = uuid.uuid4().hex
        await set_active_handoff(orch, active_handoff_id, handoff_id)
        try:
            await generate_and_save_handoff(
                orch,
                req,
                directions,
                handoff_id,
                shinychat_chat=shinychat_chat,
            )
        except Exception as e:
            if not orch.store.has(handoff_id):
                await set_active_handoff(orch, active_handoff_id, None)
            raise NotifyException(str(e)) from e

    @reactive.effect
    @reactive.event(input.handoff_close)
    async def on_close():
        await set_active_handoff(orch, active_handoff_id, None)

    @reactive.effect
    @reactive.event(input.handoff_open)
    async def on_pill_click():
        handoff_id = input.handoff_open()
        if orch.store.has(handoff_id):
            await set_active_handoff(orch, active_handoff_id, handoff_id)
            await orch.show_handoff(handoff_id)

    @reactive.effect
    @reactive.event(input.handoff_revise_text)
    async def on_revise():
        try:
            await orch.revise(active_handoff_id.get(), input.handoff_revise_text())
        except Exception as e:
            raise NotifyException(str(e)) from e
        await save_handoff_revision(shinychat_chat)

    @render.download(filename="handoff.zip")
    async def handoff_download():
        data = await orch.build_download(active_handoff_id.get())
        if data is not None:
            yield data

    def save_handoffs(values: dict[str, Any]) -> None:
        snapshot = build_handoff_snapshot(orch)
        if snapshot:
            values[HANDOFFS_BOOKMARK_KEY] = snapshot

    def restore_handoffs(values: dict[str, Any]) -> None:
        panel_close = apply_handoff_snapshot(
            orch,
            values.get(HANDOFFS_BOOKMARK_KEY),
            active_handoff_id,
        )
        if panel_close is None:
            return
        task = asyncio.create_task(panel_close)
        restore_tasks.add(task)
        task.add_done_callback(
            lambda completed: finish_handoff_restore_task(
                completed,
                restore_tasks,
            )
        )

    @session.bookmark.on_bookmark
    def on_handoff_bookmark(state: BookmarkState) -> None:
        save_handoffs(state.values)

    @session.bookmark.on_restore
    def on_handoff_bookmark_restore(state: RestoreState) -> None:
        restore_handoffs(state.values)

    @shinychat_chat.history.on_save
    def on_handoff_history_save(values: dict[str, Any]) -> None:
        save_handoffs(values)

    @shinychat_chat.history.on_restore
    def on_handoff_history_restore(values: dict[str, Any]) -> None:
        restore_handoffs(values)

File: pkg-r/R/handoff_server.R
build_handoff_snapshot <- function(orchestrator) {
  envelope <- list(
    version = 1L,
    states = lapply(
      orchestrator$snapshot(),
      handoff_state_record
    )
  )
  jsonlite::serializeJSON(envelope, digits = NA)
}

apply_handoff_snapshot <- function(
  orchestrator,
  value,
  active_handoff_id
) {
  states <- list()
  if (!is.null(value)) {
    if (
      !is.character(value) ||
        length(value) != 1L ||
        is.na(value)
    ) {
      cli::cli_abort(
        "Handoff snapshot must be a single serialized string."
      )
    }
    envelope <- jsonlite::unserializeJSON(value)
    check_plain_list(envelope, "Handoff snapshot", named = TRUE)
    check_exact_fields(
      envelope,
      c("version", "states"),
      "Handoff snapshot"
    )
    if (!identical(envelope$version, 1L)) {
      cli::cli_abort("Handoff snapshot version must be exactly 1.")
    }
    check_plain_list(
      envelope$states,
      "Handoff snapshot `states`",
      named = FALSE
    )
    states <- envelope$states
  }

  orchestrator$restore_snapshot(states)
  active_handoff_id(NULL)
  orchestrator$close_panel()
  invisible(NULL)
}

handoff_server <- function(
  input,
  output,
  session,
  chat,
  data_sources,
  executor,
  chat_module
) {
  view <- HandoffView$new(session, chat_module)
  orchestrator <- HandoffOrchestrator$new(
    chat = chat,
    data_sources = data_sources,
    executor = executor,
    view = view
  )
  active_handoff_id <- shiny::reactiveVal(
    NULL,
    label = "active_handoff_id"
  )
  save_handoffs <- function(values) {
    values$querychat_handoffs <- build_handoff_snapshot(orchestrator)
    values
  }
  restore_handoffs <- function(values) {
    apply_handoff_snapshot(
      orchestrator,
      values[["querychat_handoffs"]],
      active_handoff_id
    )
  }

  session$onBookmark(function(state) {
    state$values <- save_handoffs(state$values)
  })
  session$onRestore(function(state) {
    restore_handoffs(state$values)
  })
  chat_module$history$on_save(save_handoffs)
  chat_module$history$on_restore(restore_handoffs)

  task_handoff_id <- shiny::reactiveVal(
    NULL,
    label = "handoff_task_id"
  )
  task_operation <- shiny::reactiveVal(
    NULL,
    label = "handoff_task_operation"
  )

  recommendation_task <- shiny::ExtendedTask$new(function(items) {
    orchestrator$recommend(items)
  })
  handoff_task <- shiny::ExtendedTask$new(
    function(operation, handoff_id, value, directions) {
      task <- if (identical(operation, "generate")) {
        orchestrator$generate(value, directions, handoff_id)
      } else {
        orchestrator$revise(handoff_id, value)
      }
      promises::then(task, function(committed) {
        if (identical(operation, "revise") && identical(committed, FALSE)) {
          return(FALSE)
        }
        # The pinned shinychat history API is still in flux: the
        # `dev/querychat-pr311-history-save` branch's history object
        # exposes `on_save`/`on_restore` but no `save()` method. A missing
        # or failing history save must never mask a committed handoff by
        # rejecting this task.
        save_history <- chat_module$history$save
        if (is.function(save_history)) {
          tryCatch(
            save_history(),
            error = function(error) {
              shiny::showNotification(
                paste(
                  "The handoff was saved, but updating the chat",
                  "history failed:",
                  conditionMessage(error)
                ),
                type = "warning",
                session = session
              )
            }
          )
        }
        TRUE
      })
    }
  )

  open_handoff <- function() {
    status <- shiny::isolate(chat_module$status())
    if (status %in% c("running", "streaming")) {
      chat_module$append(
        paste(
          "Please wait for the current response to finish before",
          "preparing a handoff."
        )
      )
      return(invisible(NULL))
    }
    if (
      identical(
        shiny::isolate(recommendation_task$status()),
        "running"
      )
    ) {
      shiny::showNotification(
        "A handoff recommendation is already in progress.",
        type = "warning",
        session = session
      )
      return(invisible(NULL))
    }

    items <- orchestrator$open_modal()
    if (length(items) > 0L) {
      recommendation_task$invoke(items)
    }
    invisible(NULL)
  }

  chat_module$slash_command(
    "handoff",
    paste(
      "Prepare a shareable handoff document or webapp",
      "using the current chat context."
    ),
    open_handoff,
    echo = FALSE
  )

  output$handoff_download <- shiny::downloadHandler(
    filename = function() "handoff.zip",
    content = function(file) {
      data <- orchestrator$build_download(active_handoff_id())
      if (!is.null(data)) {
        writeBin(data, file)
      }
    }
  )

  shiny::observeEvent(
    recommendation_task$status(),
    label = "on_handoff_recommendation",
    ignoreInit = TRUE,
    {
      status <- recommendation_task$status()
      if (identical(status, "success")) {
        view$show_recommendation(recommendation_task$result())
      } else if (identical(status, "error")) {
        error_message <- tryCatch(
          {
            recommendation_task$result()
            "Unknown error"
          },
          error = function(error) conditionMessage(error)
        )
        shiny::showNotification(
          paste("Auto-recommend failed:", error_message),
          type = "error",
          duration = NULL,
          session = session
        )
        view$show_recommendation_error(error_message)
      }
    }
  )

  shiny::observeEvent(
    input$handoff_generate,
    label = "on_handoff_generate",
    {
      if (
        identical(
          shiny::isolate(handoff_task$status()),
          "running"
        )
      ) {
        shiny::showNotification(
          "A handoff is already being generated. Please wait for it to finish.",
          type = "warning",
          session = session
        )
        return()
      }

      request <- tryCatch(
        parse_handoff_generate_request(
          input$handoff_generate,
          names(handoff_registry())[[1]]
        ),
        error = function(error) {
          shiny::showNotification(
            conditionMessage(error),
            type = "error",
            duration = NULL,
            session = session
          )
          NULL
        }
      )
      if (is.null(request)) {
        return()
      }
      if (
        identical(request@type_id, "other") &&
          !nzchar(trimws(request@freeform))
      ) {
        shiny::showNotification(
          "Please enter a format name for 'Other'.",
          type = "warning",
          session = session
        )
        return()
      }

      directions <- input$handoff_directions %||% ""
      if (
        !is.character(directions) ||
          length(directions) != 1L ||
          is.na(directions)
      ) {
        directions <- ""
      }
      handoff_id <- new_handoff_id()
      active_handoff_id(handoff_id)
      task_handoff_id(handoff_id)
      task_operation("generate")
      view$set_panel_open(TRUE)
      handoff_task$invoke("generate", handoff_id, request, directions)
    }
  )

  shiny::observeEvent(
    input$handoff_revise_text,
    label = "on_handoff_revise",
    {
      if (
        identical(
          shiny::isolate(handoff_task$status()),
          "running"
        )
      ) {
        shiny::showNotification(
          "A handoff is already being generated or revised.",
          type = "warning",
          session = session
        )
        return()
      }

      handoff_id <- active_handoff_id()
      instructions <- input$handoff_revise_text
      task_handoff_id(handoff_id)
      task_operation("revise")
      handoff_task$invoke("revise", handoff_id, instructions, "")
    }
  )

  shiny::observeEvent(
    handoff_task$status(),
    label = "on_handoff_task",
    ignoreInit = TRUE,
    {
      status <- handoff_task$status()
      if (!identical(status, "error")) {
        return()
      }
      error_message <- tryCatch(
        {
          handoff_task$result()
          "Unknown error"
        },
        error = function(error) conditionMessage(error)
      )
      shiny::showNotification(
        error_message,
        type = "error",
        duration = NULL,
        session = session
      )

      if (!identical(task_operation(), "generate")) {
        return()
      }
      handoff_id <- task_handoff_id()
      committed <- tryCatch(
        isTRUE(orchestrator$show(handoff_id)),
        error = function(error) FALSE
      )
      if (!committed && identical(active_handoff_id(), handoff_id)) {
        active_handoff_id(NULL)
        view$set_panel_open(FALSE)
      }
    }
  )

  shiny::observeEvent(
    input$handoff_close,
    label = "on_handoff_close",
    {
      active_handoff_id(NULL)
      view$set_panel_open(FALSE)
    }
  )

  shiny::observeEvent(
    input$handoff_open,
    label = "on_handoff_open",
    {
      if (
        identical(
          shiny::isolate(handoff_task$status()),
          "running"
        )
      ) {
        shiny::showNotification(
          "A handoff is being generated or revised. Please wait before switching.",
          type = "warning",
          session = session
        )
        return()
      }

      handoff_id <- input$handoff_open
      if (isTRUE(orchestrator$show(handoff_id))) {
        active_handoff_id(handoff_id)
        view$set_panel_open(TRUE)
      }
    }
  )

  invisible(NULL)
}

new_handoff_id <- function() {
  paste0(
    sample(c(0:9, letters[1:6]), 32L, replace = TRUE),
    collapse = ""
  )
}
